flang/lib/Lower/OpenACC.cpp

```diff
@@ -34,50 +66,138 @@ static void genACC(Fortran::lower::AbstractConverter &absConv,
     auto &firOpBuilder = absConv.getFirOpBuilder();
     auto currentLocation = absConv.getCurrentLocation();
     llvm::ArrayRef<mlir::Type> argTy;
-    mlir::ValueRange range;
-    // Temporarly set to default 0 as operands are not generated yet.
-    llvm::SmallVector<int32_t, 2> operandSegmentSizes(/*Size=*/7,
-                                                      /*Value=*/0);
-    auto loopOp =
-        firOpBuilder.create<mlir::acc::LoopOp>(currentLocation, argTy, range);
-    loopOp.setAttr(mlir::acc::LoopOp::getOperandSegmentSizeAttr(),
-                   firOpBuilder.getI32VectorAttr(operandSegmentSizes));
+
+    // Add attribute extracted from clauses.
+    const auto &accClauseList =
+        std::get<Fortran::parser::AccClauseList>(beginLoopDirective.t);
+
+    mlir::Value workerNum = nullptr;
+    mlir::Value vectorLength = nullptr;
+    mlir::Value gangNum = nullptr;
+    mlir::Value gangStatic = nullptr;
+    std::int32_t tileOperands = 0;
+    std::int32_t privateOperands = 0;
+    std::int32_t reductionOperands = 0;
+    std::int64_t executionMapping = mlir::acc::OpenACCExecMapping::NONE;
+    SmallVector<Value, 8> operands;
+
+    // Lower clauses values mapped to operands.
+    for (const auto &clause : accClauseList.v) {
+      if (const auto *gangClause =
+              std::get_if<Fortran::parser::AccClause::Gang>(&clause.u)) {
+        if (gangClause->v) {
+          const Fortran::parser::AccGangArgument &x = *gangClause->v;
+          if (const auto &gangNumValue =
+                  std::get<std::optional<Fortran::parser::ScalarIntExpr>>(
+                      x.t)) {
+            gangNum = absConv.genExprValue(
+                *Fortran::semantics::GetExpr(gangNumValue.value()));
+            operands.push_back(gangNum);
+          }
+          if (const auto &gangStaticValue =
+                  std::get<std::optional<Fortran::parser::AccSizeExpr>>(x.t)) {
+            const auto &expr =
+                std::get<std::optional<Fortran::parser::ScalarIntExpr>>(
+                    gangStaticValue.value().t);
+            if (expr.has_value()) {
+              gangStatic = absConv.genExprValue(
+                  *Fortran::semantics::GetExpr(expr.value()));
+            } else {
+              // * was passed as value.
+              gangStatic = firOpBuilder.createIntegerConstant(
+                  currentLocation, firOpBuilder.getIntegerType(32), -1);
+            }
+            operands.push_back(gangStatic);
+          }
+        }
+        executionMapping |= mlir::acc::OpenACCExecMapping::GANG;
+      } else if (const auto *workerClause =
+                     std::get_if<Fortran::parser::AccClause::Worker>(
+                         &clause.u)) {
+        if (workerClause->v) {
+          workerNum = absConv.genExprValue(
+              *Fortran::semantics::GetExpr(workerClause->v.value()));
```

Comment:
Convention note: if the `std::optional` has already been tested and found not empty, the `*` operator is used to avoid further checks.